pkg/controller/clone-controller.go

```diff
@@ -102,162 +185,172 @@ func (c *CloneController) Run(threadiness int, stopCh <-chan struct{}) error {
+	}
+	if targetPod != nil && !metav1.IsControlledBy(sourcePod, pvc) {
+		return errors.Errorf("found pod %s/%s not owned by pvc %s/%s", targetPod.Namespace, targetPod.Name, pvc.Namespace, pvc.Name)
 	}
+	
+	// expectations prevent us from creating multiple pods. An expectation forces
+	// us to observe a pod's creation in the cache.
+	needsSync := c.podExpectations.SatisfiedExpectations(pvcKey)
 
-	// update pvc with 'AnnCloningPods' annotation to indicate cloning is in process
-	anno := map[string]string{AnnCloningPods: "exist"}
-	var lab map[string]string
-	pvc, err = updatePVC(c.clientset, pvc, anno, lab)
-	if err != nil {
-		return errors.WithMessage(err, "could not update pvc %q annotation and/or label")
+	// make sure not to reprocess a PVC that has already completed successfully,
+	// even if the pod no longer exists
+	previousPhase, exists := pvc.ObjectMeta.Annotations[AnnClonePodPhase]
+	if exists && (previousPhase == string(v1.PodSucceeded)) {
+		needsSync = false
 	}
 
-	//create random string to be used for pod labeling and hostpath name
-	generatedLabelStr := util.RandAlphaNum(GENERATED_CLONING_LABEL_LEN)
-	//create the source pod
-	pod, err := CreateCloneSourcePod(c.clientset, c.cloneImage, c.verbose, c.pullPolicy, cr, pvc, generatedLabelStr)
-	if err != nil {
-		//TODO: remove annotation AnnCloningPods from pvc as pod failed to run
-		return err
+	if needsSync && (sourcePod == nil || targetPod == nil){
+		//create random string to be used for pod labeling and hostpath name
+		generatedLabelStr := util.RandAlphaNum(GENERATED_CLONING_LABEL_LEN)
+		if sourcePod == nil {
+			cr, err := getCloneRequestPVC(pvc)
+			if err != nil {
+				return err
+			}
+			// all checks passed, let's create the cloner pods!
+			c.expectPodCreate(pvcKey)
+			//create the source pod
+			sourcePod, err = CreateCloneSourcePod(c.clientset, c.cloneImage, c.verbose, c.pullPolicy, cr, pvc, generatedLabelStr)
+			if err != nil {
+				c.observePodCreate(pvcKey)
+				return err
+			}
+		}
+		if targetPod == nil {
+			c.expectPodCreate(pvcKey)
+			//create the target pod
+			targetPod, err = CreateCloneTargetPod(c.clientset, c.cloneImage, c.verbose, c.pullPolicy, pvc, generatedLabelStr, sourcePod.ObjectMeta.Namespace)
+			if err != nil {
+				c.observePodCreate(pvcKey)
+				return err
+			}
+		}
+		return nil
 	}
 
-	//create the target pod
-	_, err = CreateCloneTargetPod(c.clientset, c.cloneImage, c.verbose, c.pullPolicy, pvc, generatedLabelStr, pod.ObjectMeta.Namespace)
-	if err != nil {
-		//TODO: remove annotation AnnCloningPods from pvc as pod failed to run
-		return err
+	// update pvc with cloner pod name and optional cdi label
+	//we update the target PVC according to the target pod. Only the target pods indicates the real status of the cloning.
+	if targetPod.Spec.Affinity.PodAffinity != nil { // this is the target clone pod
+		anno[AnnClonePodPhase] = string(targetPod.Status.Phase)
```

Comment:
maybe only overwrite the phase if the phase wasn't previously succeeded. In general, It should be possible for both the target and src pods to disappear and not cause any issues once the phase hits succeeded.

Reply:
This was copied from the importer. I think that it's a wrong wording. should be phase instead previousPhase. I will change the code as this is confusing.